build/xtask/src/dist.rs

```diff
@@ -2452,69 +2452,125 @@ fn write_elf(
     kentry: AbiSize,
     cfg: &PackageConfig,
     out: &Path,
+) -> Result<()> {
+    if cfg.arch_consts.objcopy_target.starts_with("elf64") {
+        println!("Building ELF64");
+        write_elf64(sections, kentry, cfg, out)?;
+    } else {
+        println!("Building ELF32");
+        write_elf32(sections, kentry, cfg, out)?;
+    }
+
+    Ok(())
+}
+
+fn write_elf32(
+    sections: &BTreeMap<AbiSize, LoadSegment>,
+    kentry: AbiSize,
+    cfg: &PackageConfig,
+    out: &Path,
 ) -> Result<()> {
     use goblin::container::{Container, Ctx, Endian};
-    use goblin::elf::header::{Header, ET_EXEC};
-    use goblin::elf::program_header::{PF_R, PF_W, PF_X, PT_LOAD};
-    use goblin::elf::section_header::{
-        SHF_ALLOC, SHF_EXECINSTR, SHT_NULL, SHT_PROGBITS, SHT_STRTAB,
+    use goblin::elf32::header::{Header, ET_EXEC};
+    use goblin::elf32::program_header::ProgramHeader;
+    use goblin::elf32::program_header::{PF_R, PF_W, PF_X, PT_LOAD};
+    use goblin::elf32::section_header::SectionHeader;
+    use goblin::elf32::section_header::{
+        SHF_ALLOC, SHF_EXECINSTR, SHT_PROGBITS, SHT_STRTAB,
     };
-    use goblin::elf::{ProgramHeader, SectionHeader};
     use scroll::Pwrite;
 
-    // 'Big' Containers are Goblin for ELF64. 'Little' are ELF32.
-    let ctx = Ctx::new(
-        if cfg.arch_consts.objcopy_target.starts_with("elf64") {
-            Container::Big
-        } else {
-            Container::Little
-        },
-        Endian::Little,
-    );
+    let ctx = Ctx::new(Container::Little, Endian::Little);
+
+    let mut sections_base_address: AbiSize = kentry;
+    let mut sections_length: u64 = 0;
+
+    for candidate_section in sections {
+        if candidate_section.1.data.len() > 0 {
+            if *candidate_section.0 < sections_base_address {
+                sections_base_address = *candidate_section.0;
+            }
```

Comment:
Is this handling the condition where some task is placed before the kernel in the layout?

Reply:
Yes, it's a preventative guard. I can't say I've seen an existing target that does this, but I also couldn't identify what would stop someone from putting something in advance of the kernel.